### PR TITLE
feat(protocol-designer): off-deck labware visual representation

### PR DESCRIPTION
**Pull request by jerader**
closes RAUT-696

# Overview

Now, there is a way to see visually which labware is off deck in PD at any given part of the protocol. And you can add off deck labware initially during the starting deck state.

# Test Plan

sandbox: http://sandbox.designer.opentrons.com/pd_off-deck-labware-in-setup/

- go to the setting and turn on the feature flag for enabling off deck visuals and multi tiprack support (keeping this on a new feature flag since this + multi tiprack are currently supported on the app and therefore, these features are very likely to be in PD 7.1).
- create a protocol for the Flex, go to the deck setup, click on the `edit off deck labware` button. The slideout should appear. Add a labware. Make sure the options to add nickname, add liquids, duplicate, delete are there and work.
- add a step to the protocol, click on the edit off deck labware button. Make sure the only option available is to view liquids. Make sure the button in the slideout is disabled.
- add a move labware step to move an off deck labware to the deck. Make sure the labware entity disappears from the slideout.
- repeat steps for OT-2 protocol
- export a protocol and in the json file, make sure the `loadlabware` that you loaded off deck initially says the location is off deck.

# Changelog

- make a slideout component in PD (its basically a copy of the app slideout)
- create `EditOffDeckLabware` 
- edit `browseLabware` and `NameThisLabware`'s prop to take in just the labware entity
- create new `OffDeckLabwareSlidoeut` and `OffDeckLabwareButton`
- some styling z-index changes

# Review requests

see test plan

# Risk assessment

low

### CODE REVIEW

**Comment by jerader** [comment]
- [x] fix merge conflicts
- [x] fix bug where you can't move a labware off deck if there is already a labware off deck.
- [x] need to really thoroughly smoke test this with the feature flag turned off to make sure nothing breaks, especially if I merge this before the last release candidate for PD 7.0.0 -> the final release candidate for 7.0 is hopefully cut already 🤞

**Comment by koji** [comment]
@jerader 
the option name is `Allow all tip rack options` ?

**Comment by jerader** [comment]
> @jerader the option name is `Allow all tip rack options` ?

@koji, oh you have to turn on the `Off deck visualization and multi tiprack support` feature flag which is under dev ffs, so you have to run `enablePrereleaseMode()` in console and refresh to reveal them. They'll be in the setting page underneath the other feature flags

File: protocol-designer/src/localization/en/button.json [hunk 1]
```diff
@@ -1,5 +1,6 @@
 {
   "add_step": "+ Add Step",
+  "add_off_deck": "+ Off deck labware",
```
Comment (ecormany) [inline protocol-designer/src/localization/en/button.json]: Can we hyphenate `Off-deck labware` here and throughout?

See https://docs.opentrons.com/v2/moving_labware.html#the-off-deck-location for some more examples in context.

File: protocol-designer/src/localization/en/tooltip.json [hunk 2]
```diff
@@ -3,6 +3,7 @@
   "advanced_settings": "Advanced Settings",
 
   "disabled_step_creation": "New steps cannot be added in Batch Edit mode.",
+  "disabled_off_deck": "Off deck labware cannot be modified unless on starting deck state.",
```
Comment (ecormany) [inline protocol-designer/src/localization/en/tooltip.json]: When does this trigger? Even on-deck labware can't be modified at later steps, right?

**Review by ecormany — COMMENTED** [review]
Coupla copy comments. Mostly about hyphenating `off-deck`.

File: protocol-designer/src/localization/en/tooltip.json (hunk 2 above)
Comment (jerader) [reply protocol-designer/src/localization/en/tooltip.json]: This is the tooltip for the slideout button for specifically adding off deck labware.
Screen Shot 2023-09-25 at 11 42 39 AM

Comment (jerader) [reply protocol-designer/src/localization/en/tooltip.json]: But yes, all labware can not be modified if its not the starting deck state.

File: protocol-designer/src/localization/en/button.json (hunk 1 above)
Comment (jerader) [reply protocol-designer/src/localization/en/button.json]: Thanks! this is nice to know - i definitely know a few other instances where it needs to be hyphenated.

**Comment by koji** [comment]
When added more than 8 labwares, the overlay items that the latest labware showed were different from the above labwares.
But after mouse over a labware, the displayed item became the same.

https://github.com/Opentrons/opentrons/assets/474225/ec0720fa-d88e-49a1-847c-6e4db69dc7e7

**Comment by koji** [comment]
Question

When move labware from deck to off-deck, the labware is added to the slideout and it is placed in the top.
Is that expected behavior?

I was wondering, if we followed the slideout behavior, maybe adding that labware to the bottom would be a consistency behavior?

**Comment by jerader** [comment]
> Question
> 
> When move labware from deck to off-deck, the labware is added to the slideout and it is placed in the top. Is that expected behavior?
> 
> I was wondering, if we followed the slideout behavior, maybe adding that labware to the bottom would be a consistency behavior?

@koji good question, I didn't notice that this is different behavior from the app. I think it makes sense to make the behaviors the same. I'll make a follow up ticket for it though because it is a bit complicated (I think) to tell when the labware has been moved off deck.

File: protocol-designer/src/atoms/Slideout.tsx [hunk 3]
```diff
@@ -0,0 +1,227 @@
...
+  return (
+    <>
+      <Overlay
+        onClick={handleClose}
+        css={`
+          ${isExpanded ?? false ? OVERLAY_IN_STYLE : overlayOutStyle}
+        `}
+        backgroundColor={COLORS.transparent}
+        height="92.5%"
+        marginTop="3rem"
```
Comment (koji) [inline protocol-designer/src/atoms/Slideout.tsx]:
```suggestion
        marginTop={SPACING.spacing48}
```

File: protocol-designer/src/atoms/Slideout.tsx [hunk 4]
```diff
@@ -0,0 +1,227 @@
...
+      <Box
+        css={isExpanded ?? false ? EXPANDED_STYLE : collapsedStyle}
+        cursor="auto"
+        position={POSITION_FIXED}
+        right="0"
+        top="0"
+        backgroundColor={COLORS.white}
+        boxShadow="0px 3px 6px rgba(0, 0, 0, 0.23)"
+        height="92.5%"
+        marginTop="3rem"
```
Comment (koji) [inline protocol-designer/src/atoms/Slideout.tsx]:
```suggestion
        marginTop={SPACING.spacing48}
```

File: protocol-designer/src/atoms/Slideout.tsx [hunk 5]
```diff
@@ -0,0 +1,227 @@
...
+              <Flex alignItems={ALIGN_CENTER}>
+                <Btn
+                  size="1.5rem"
+                  onClick={handleClose}
+                  aria-label="exit"
+                  data-testid={`Slideout_icon_close_${
+                    typeof title === 'string' ? title : ''
+                  }`}
+                >
+                  <Icon name="close" css={CLOSE_ICON_STYLE} />
```
Comment (koji) [inline protocol-designer/src/atoms/Slideout.tsx:189]: nit

maybe the size of the icon a little bit small when compare to the slideout title?

**Comment by jerader** [comment]
> When added more than 8 labwares, the overlay items that the latest labware showed were different from the above labwares. But after mouse over a labware, the displayed item became the same.
> 

@koji looking into this - this is actually expected behavior. When you click to add a new labware that is not duplicating, it displays the other overlay first to allow you the opportunity to add a nickname to the labware. If you are duplicating, you can't nick name the labware since you are just duplicating the other labware's nickname/display name. This behavior also matches the labware on the deck too.

File: protocol-designer/src/atoms/Slideout.tsx (hunk 5 above)
Comment (jerader) [reply protocol-designer/src/atoms/Slideout.tsx:189]: i was just thinking about this too, I will increase the size, thanks!

File: protocol-designer/src/components/DeckSetup/LabwareOverlays/EditLabwareOffDeck.tsx [hunk 6]
```diff
@@ -0,0 +1,144 @@
...
+          z-index: 1;
+          bottom: 0;
+          position: ${POSITION_ABSOLUTE};
+          width: 127.76px;
+          height: 85.45px;
+          opacity: 0;
+          &:hover {
+            opacity: 1;
+          }
+        `}
```
Comment (koji) [inline protocol-designer/src/components/DeckSetup/LabwareOverlays/EditLabwareOffDeck.tsx]: nit
maybe create a style constant.

File: protocol-designer/src/components/DeckSetup/LabwareOverlays/EditLabwareOffDeck.tsx [hunk 7]
```diff
@@ -0,0 +1,144 @@
...
+          bottom: 0;
+          font-size: 0.7rem;
+          position: ${POSITION_ABSOLUTE};
+          width: 127.76px;
+          height: 85.45px;
+          opacity: 0;
+          &:hover {
+            opacity: 1;
+          }
+        `}
```
Comment (koji) [inline protocol-designer/src/components/DeckSetup/LabwareOverlays/EditLabwareOffDeck.tsx]: This one is long so I recommend to create a style const for this.

`border-radius: BORDERS.borderRadiusSize4`

**Comment by jerader** [comment]
to address the off-deck labware slideout populating from latest added off-deck labware. https://opentrons.atlassian.net/browse/RAUT-757

**Comment by koji** [comment]
Aluminum Flat Bottom Plate's overlay position is a little bit off.

https://github.com/Opentrons/opentrons/assets/474225/42e9c8eb-07d4-4460-886c-bcdc4426911c


Seems that the svg in the modal is a little bit off? (not a big issue)
Screenshot 2023-09-25 at 12 36 49 PM

**Review by koji — APPROVED** [review]
left a few comments.
I followed the test plan and the sandbox worked as expected

**Comment by jerader** [comment]
> Aluminum Flat Bottom Plate's overlay position is a little bit off.
> 

oops @koji good catch! the adapters are positioned a bit differently using the `cornerOffsetFromSlot`. I bet all the adapters are slightly off like this. I'll fix it. Thanks!

**Pull request merged**